### PR TITLE
OTA: Design document

### PR DESCRIPTION
**Pull request by lightsofapollo**
This is what I came up with in terms of a reasonable good design that we can implement very quickly. I estimate this at around 1-2 weeks of work (personal estimation for myself only) including full suite of tests, documentation, etc... 

Please speak up and ask questions if anything in unclear. My objective is anyone can read the design document then the tests and be comfortable jumping into the OTA code. Fabrice made a good point that we _should_ use an existing off the shelf solution for updates. I don't disagree but I am not sure how urgently we need production ready updates and I would prefer to have more control over this system while we decide what is most important.

### CODE REVIEW

**Comment by lightsofapollo** [comment]
The `WAL` life cycles are not very clear in the document. I will more clearly lay those out as I dig deeper into how the code will work. I will have no problem expanding this now but the end result may be slightly different than what would be in the document.

File: ota/README.md [hunk 1]
```diff
@@ -0,0 +1,69 @@
+# Gateway OTA updater.
+
+This folder contains the package which deals with OTA updates for the gateway. The OTA process includes downloading, extracting and installing : node, build assets, native dependencies and `node_modules`. OTA does not update any of the underlying system libraries.
```
Comment (lightsofapollo) [inline ota/README.md]: NOTE: Like the rest of the document this is the end goal... I like splitting this kind of work over multiple pull requests.

File: ota/README.md [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+
+Write ahead log (WAL): Must use "atomic" append only writes. TODO: find link describing how linux kernel does this.
+
+### `update.json`.
+
+This is used for high level illustration we may need additional fields.
+
+`version`: A monotonically increasing number. We use a number rather than a UUID to allow for the underlying update channel to be "eventually consistent" where we may sometimes find a number lower than the current update.
+
+`contentAddressabeURL`: The URL which points to the new update in the format of `<hash of tar>.tar.gz`. By using a hash to address the URL it makes it more difficult to pull off a man in the middle attack (which is still possible even over SSL).
```
Comment (hobinjk) [inline ota/README.md]: nit: `contentAddressabeURL` -> `contentAddressableURL`.

File: ota/README.md [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+  "version": 1,
+  "contentAddressableURL": "5dde896887f6754c9b15bfe3a441ae4806df2fde94001311e08bf110622e0bbe.tar.gz"
+}
+```
+
+### High level operations
+
+#### Downloading
+
+`currentVersion`: The currently installed version running on the gateway. Note this version may or may not be the same as `package.json.version`.
```
Comment (hobinjk) [inline ota/README.md]: Under what conditions would they diverge?

Comment (lightsofapollo) [inline ota/README.md]: Excellent question. My use of "currentVersion" is misleading what we really want here I think is "timestamp". For example I would like to release at least once a day and that (I believe) is very inconvenient to do a full version bump. So we will have something like `package.json.version` 0.3 and in this json file we will have "timestamp" (used to figure out when to download an update) and thinking further I also want to add a "revision" field so we can tell exactly what version of the code is deployed.

**Comment by lightsofapollo** [comment]
@dhylands I made a mistake not describing the directory structure and how it relates to the "WAL". I will attempt to explain my intent here and then update the document and we can do another review pass to make sure it is clear.

In the simplest implementation (which is what we will start with) your thinking is exactly right we have a "A" directory and a "B" directory. I gave it some more thought and while I like this concept of "A/B" what I really meant was "We can have N versions of the gateway installed at the same time. We can switch between them and they may exist across multiple channels". I want to implement what I think are the underlying pieces to this without implementing more than we need over the next quarter.

My use of the term "WAL" comes from context of databases. Even among databases how the "WAL" is used and how it is actually implemented vary. "WAL" in this case (maybe I need another name) is a place where the system logs update operations before executing them and in the case of failure TBD system can recover from the WAL or perhaps just output a better error. Initially "WAL" is just a fancy way of saying : "A log of updates, their version, revision and channel and the status of the update". Note that it is important to me that this works in multi threaded situations as multiple systems may (or may not) write to the WAL. I will include my source material in the doc.

That log is useful for doing debugging (i.e. WTF why did my update break updater and entered us into a loop) and possibly useful in the future for some more advanced mechanism which can read the WAL and decide what to do (i.e. if the update does not apply maybe try again in N amount o time). My intent here is lets store this data early and figure out how it is useful along the way. It could be very useful to also have a quick way to opt into sharing this log with us (Mozilla) so we can potentially diagnose updater issues in the field (without sharing potentially more sensitive logs).

**Comment by dhylands** [comment]
Databases work by essentially having to keep the previous version of the data and the new version and rolling back to the previous version if the new version doesn't get written. They don't have the additional constraint of having to be linear.

When writing to a log file, there are 2 very distinct layers that need to be considered. The first is the "file" layer, and if the file is opened in append mode, then it will ensure that writes below a certain size are atomic. This just means that if thread/process A writes to the file at the same time as thread/process B, then their writes will not be intermingled.

The second layer is the actual backing store (i.e. hard drive or sdcard) and ensuring that the file data has actually been persisted. But there isn't any way of guaranteeing that a write from the "file" layer that crosses physical block boundaries will get completely written to disk. If the power is pulled there will always be situations where only a portion of the write gets to the disk. This is why the databases have to keep old and new.

So I'm primarily pointing out that keeping it linear like a log file has disadvantages as far as data integrity is concerned.

And I want to make sure that you understand that being atomic and maintaining data integrity are 2 different concepts.

**Comment by lightsofapollo** [comment]
@dhylands Valid concerns let me try to address them.

>Databases work by essentially having to keep the previous version of the data and the new version and rolling back to the previous version if the new version doesn't get written. They don't have the additional constraint of having to be linear.

The way I used "databases" is vague. I am talking about databases designed for [CAP](https://en.wikipedia.org/wiki/CAP_theorem) , distributed systems and which primarily were engineered for the cloud environment in which machines fail regularly and that is simply part of the engineering problem. It's been years since I actually read the code but https://docs.mongodb.com/manual/core/write-performance/ talks about the write ahead log and how it is used and this is where I am drawing my context from originally. I am realizing that this term could mean a number of things and I could use a better name.

>When writing to a log file, there are 2 very distinct layers that need to be considered. The first is the "file" layer, and if the file is opened in append mode, then it will ensure that writes below a certain size are atomic. This just means that if thread/process A writes to the file at the same time as thread/process B, then their writes will not be intermingled.

Correct ^ this is also my understand (I am talking about O_APPEND) I once had a great piece of reference material which talked both about how writes would not be intermingled but also the maximum size of write which can be atomic. This IIRC also varies upon the underling file system. I believe ext3, ex4, brtfs, zfs all behave similarly with varying maximum write sizes (NFS for example has different behavior). The entries in the WAL must be relatively small and have checks to ensure we don't write more that whatever the maximum is. Some quick reference material I could dig up (write size)[http://ar.to/notes/posix#pipe-buf] and (append)[http://pubs.opengroup.org/onlinepubs/9699919799/functions/write.html]. When we decide what our final "OS" looks like I plan on verifying that the maximum write is what I think it is.

>The second layer is the actual backing store (i.e. hard drive or sdcard) and ensuring that the file data has actually been persisted. But there isn't any way of guaranteeing that a write from the "file" layer that crosses physical block boundaries will get completely written to disk. If the power is pulled there will always be situations where only a portion of the write gets to the disk. This is why the databases have to keep old and new.
>And I want to make sure that you understand that being atomic and maintaining data integrity are 2 different concepts.

I definitely understand they are two separate issues (this is part of the reason you don't see flushing to disk or fsync in this document). From my understanding _when_ and _how_ to attempt to "flush" (Is this the right term?) to disk from userspace is a complex problem and I don't have a concrete answer as to why we _should_ or _should not_ fsync/flush yet as the log is simply "best effort". 

>So I'm primarily pointing out that keeping it linear like a log file has disadvantages as far as data integrity is concerned.

True, Often in the database systems I am familiar with use this is a last resort and then the failed node must do some consensus operation with other nodes. Having a best effort log gives us useful information that may or may not be useful in a unexpected shutdown/failure situation. This is usually highly configurable because different systems have different "write concerns".

Can you in the future use the github review feature ? This is very useful as you can split your concerns across multiple comments which individually can be responded to. This makes it clearer to the reviewer/reviewee if a comment was missed or ignored. I try to respond to each of these comments (even if it's just a +1) so my reviewer knows I read it and either took action or responded to the comment. My bias is to answer the questions that I can answer quickly then come back and respond to the more detailed comments later on.

**Pull request closed without merging**